File: 93-restore-ip-addresses/Sony/93. Restore IP Addresses.py
class Solution(object):
    def restoreIpAddresses(self, s):
        """
        :type s: str
        :rtype: List[str]
        """
        def parseInt(string):
            int_val = int(string)

            if len(str(int_val)) != len(string):
                return 555 #a invalid value
            else:
                return int_val

        def genSegment(seg_left, in_str):
            result = list()
            if seg_left * 3 < len(in_str) or len(in_str) == 0:
                return []

            if seg_left > 1:
                for i in xrange(1, 4):
                    prefix = in_str[0:i]
                    suffix = in_str[i:]
                    int_val = parseInt(prefix)
                    if int_val > 255:
                        continue

                    ret = genSegment(seg_left - 1, suffix)
                    for sub_str in ret:
                        result.append(prefix + '.' + sub_str)
            elif parseInt(in_str) <= 255:
                result.append(in_str)

            return result

        return genSegment(4, s)

if __name__ == '__main__':
    sol = Solution()

    ip = "0000"

    print sol.restoreIpAddresses(ip)

File: 93-restore-ip-addresses/leochang/93-restore-ip-addresses.cpp
class Solution {
public:
	vector<string> restoreIpAddresses(string s) {
		vector<string> result;
		int len = s.size();

		for (int i = 1; i < 4; ++i) {
			for (int j = 1; j < 4; ++j) {
				for (int k = 1; k < 4; ++k) {
					if (i + j + k >= len) continue;
					string s1 = s.substr(0, i);
					string s2 = s.substr(i, j);
					string s3 = s.substr(i+j, k);
					string s4 = s.substr(i+j+k, len-i-j-k);
					if (isValid(s1) && isValid(s2) &&
							isValid(s3) && isValid(s4)) {
						string sol = s1 + "."+ s2 + "."+ s3 + "."+ s4;
						result.push_back(sol);
					}
				}
			}
		}
		return result;
	}
	bool isValid(string s){
		if (s.size() > 3 || s.size() == 0 || stoi(s) > 255 || (s.front()=='0' && s.size() > 1)) return false;
		return true;
	}
};
